### PR TITLE
lightning, membuf: limit the maximum memory size allocated from pool

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/48420

Problem Summary:

### What changed and how does it work?

- Add a limiter to membuf.Pool which blocks the caller when it has allocated too many memory.

- Concurrent reader uses multiple small buffers rather than a large buffer to reduce overhead, because membuf.Pool can only allocate fixed size memory blocks

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

avoid OOM in 300G test

img_v3_025u_9c56b0e3-3396-48a1-a568-ea9cfd15f34g

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by wjhuang2016 — APPROVED** [review]

File: br/pkg/lightning/backend/external/byte_reader.go [hunk 1]
```diff
@@ -195,42 +203,67 @@ func (r *byteReader) switchToConcurrentReader() error {
...
-func (r *byteReader) next(n int) []byte {
-	end := min(r.curBufOffset+n, len(r.curBuf))
-	ret := r.curBuf[r.curBufOffset:end]
-	r.curBufOffset += len(ret)
-	return ret
+func (r *byteReader) next(n int) (int, [][]byte) {
+	retCnt := 0
+	// TODO(lance6716): heap escape performance?
+	ret := make([][]byte, 0, len(r.curBuf)-r.curBufIdx+1)
+	for r.curBufIdx < len(r.curBuf) && n > 0 {
```
Comment (D3Hunter) [inline br/pkg/lightning/backend/external/byte_reader.go:247]: more and more complex memory operation code😅

File: br/pkg/membuf/limiter.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+		l.mu.Unlock()
+		return
+	}
+
+	waitCh := make(chan struct{})
+	l.waitNums = append(l.waitNums, n)
+	l.waitChs = append(l.waitChs, waitCh)
+	l.mu.Unlock()
+
+	<-waitCh
```
Comment (D3Hunter) [inline br/pkg/membuf/limiter.go:50]: how can we avoid deadlock on this?

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

File: br/pkg/membuf/limiter.go (hunk 2 above)
Comment (lance6716) [reply br/pkg/membuf/limiter.go:50]: caller need to ensure every one task which acquires memory should not exceed the memory pool limit 🤔 I have no other good ideas now.

**Pull request merged**